File: packages/quran-engine-swift/Sources/QuranEngine/Audio.swift
import Foundation

/// Zero-pad an integer to 3 digits ("001", "057", "114"). Mirrors `zeroPad3`.
public func zeroPad3(_ n: Int) -> String {
    String(format: "%03d", n)
}

public enum AudioError: Error, Equatable {
    case surahOutOfRange(Int)
    case noSurahFeed(String)
}

/// Full-surah recitation URL: `surahLink + zeroPad3(surah) + ".mp3"`. Mirrors `surahAudioUrl`.
public func surahAudioURL(_ reciter: Reciter, _ surahNumber: Int) throws -> String {
    guard surahNumber >= 1 && surahNumber <= 114 else { throw AudioError.surahOutOfRange(surahNumber) }
    guard !reciter.surahLink.isEmpty else { throw AudioError.noSurahFeed(reciter.name) }
    return "\(reciter.surahLink)\(zeroPad3(surahNumber)).mp3"
}

/// Ayah-by-ayah recitation URL. Requires the global ayah number (1..6236). Mirrors `ayahAudioUrl`.
public func ayahAudioURL(_ reciter: Reciter, _ globalAyahNumber: Int) -> String {
    "https://cdn.islamic.network/quran/audio/\(reciter.ayahBitrate)/\(reciter.ayahIdentifier)/\(globalAyahNumber).mp3"
}

private let minshawiFallbackName = "Muhammad Al-Minshawi (Murattal)"

/// True if this reciter falls back to Minshawi for individual-ayah audio.
public func defaultsToMinshawi(_ reciter: Reciter) -> Bool {
    reciter.ayahIdentifier.contains("minshawi") && !reciter.name.contains("Minshawi")
}

/// Display name to show while ayah audio plays (honest about the fallback).
public func ayahNowPlayingName(_ reciter: Reciter) -> String {
    if defaultsToMinshawi(reciter) { return minshawiFallbackName }
    if let q = reciter.qiraah { return "\(reciter.name) (\(q))" }
    return reciter.name
}

/// Reciter directory. Mirrors `Reciters` in `audio.js`. Sorted by name on construction.
public final class Reciters {
    public let list: [Reciter]
    private let byIdMap: [String: Reciter]

    public init(_ list: [Reciter]) {
        self.list = list.sorted { $0.name.localizedCompare($1.name) == .orderedAscending }
        var m = [String: Reciter]()
        for r in self.list { m[r.id] = r }
        self.byIdMap = m
    }

    public func all() -> [Reciter] { list }

    public func byId(_ id: String) -> Reciter? { byIdMap[id] }

    /// Reciters that have a full-surah feed.
    public func withSurahFeed() -> [Reciter] {
        list.filter { !$0.surahLink.isEmpty && !$0.surahLink.hasSuffix(".mp3") }
    }

    /// Reciters for a given riwayah label (`nil`/"hafs" => the default Hafs feeds).
    public func byQiraah(_ qiraah: String?) -> [Reciter] {
        guard let q = qiraah, q.lowercased() != "hafs" else { return list.filter { $0.qiraah == nil } }
        return list.filter { $0.qiraah == q }
    }

    /// Distinct riwayah labels available (excluding default Hafs).
    public func qiraat() -> [String] {
        var seen = Set<String>()
        var out = [String]()
        for r in list { if let q = r.qiraah, seen.insert(q).inserted { out.append(q) } }
        return out
    }
}

File: packages/quran-engine-go/audio.go
package quranengine

import (
	"fmt"
	"sort"
	"strings"
)

// pad3 zero-pads to three digits: 1 -> "001", 57 -> "057", 114 -> "114".
func pad3(n int) string { return fmt.Sprintf("%03d", n) }

// Reciters returns all reciters sorted by name (matching the JS port's display order).
func (e *Engine) Reciters() []Reciter {
	out := make([]Reciter, len(e.reciters))
	copy(out, e.reciters)
	sort.SliceStable(out, func(i, j int) bool { return out[i].Name < out[j].Name })
	return out
}

// ReciterByID returns the reciter with the given id, or nil if not found.
func (e *Engine) ReciterByID(id string) *Reciter {
	for i := range e.reciters {
		if e.reciters[i].ID == id {
			return &e.reciters[i]
		}
	}
	return nil
}

// SurahAudioURL builds the full-surah recitation URL:
//
//	reciter.SurahLink + pad3(surah) + ".mp3"
func SurahAudioURL(r *Reciter, surahNumber int) (string, error) {
	if surahNumber < 1 || surahNumber > 114 {
		return "", fmt.Errorf("surah out of range: %d", surahNumber)
	}
	if r.SurahLink == "" {
		return "", fmt.Errorf("reciter %q has no full-surah feed", r.Name)
	}
	return r.SurahLink + pad3(surahNumber) + ".mp3", nil
}

// AyahAudioURL builds the per-ayah recitation URL. globalAyahNumber is the
// 1..6236 value from Engine.GlobalAyahNumber.
//
//	https://cdn.islamic.network/quran/audio/{bitrate}/{identifier}/{globalAyah}.mp3
func AyahAudioURL(r *Reciter, globalAyahNumber int) string {
	return fmt.Sprintf("https://cdn.islamic.network/quran/audio/%s/%s/%d.mp3",
		r.AyahBitrate, r.AyahIdentifier, globalAyahNumber)
}

const minshawiFallbackName = "Muhammad Al-Minshawi (Murattal)"

// DefaultsToMinshawi reports whether a reciter falls back to Minshawi for ayah audio.
func DefaultsToMinshawi(r *Reciter) bool {
	return strings.Contains(r.AyahIdentifier, "minshawi") && !strings.Contains(r.Name, "Minshawi")
}

// AyahNowPlayingName returns the display name to show while ayah audio plays.
func AyahNowPlayingName(r *Reciter) string {
	if DefaultsToMinshawi(r) {
		return minshawiFallbackName
	}
	if r.Qiraah != "" {
		return fmt.Sprintf("%s (%s)", r.Name, r.Qiraah)
	}
	return r.Name
}
